internal/provider/resource_access_model.go

```diff
@@ -5,186 +5,300 @@ import (
+		)
+		return
+	}
+	a.client = client
+}
 
-	model := d.Get("model").(string)
-	access := d.Get("access").(string)
-	usersInterface := d.Get("users").([]interface{})
-	users := make([]string, len(usersInterface))
-	for i, v := range usersInterface {
-		users[i] = v.(string)
+func (a *accessModelResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	// Check first if the client is configured
+	if a.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "create")
+		return
 	}
+	var plan accessModelResourceModel
 
-	uuid, err := client.Models.ResolveModelUUID(model)
-	if err != nil {
-		return diag.FromErr(err)
+	// Read Terraform configuration from the request into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
+	// Get the users
+	users := convertRawUsers(plan.Users.Elements())
+
+	modelNameStr := plan.Model.ValueString()
+	// Get the modelUUID to call Models.GrantModel
+	uuid, err := a.client.Models.ResolveModelUUID(modelNameStr)
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to get model uuid, got error: %s", err))
+		return
+	}
 	modelUUIDs := []string{uuid}
 
+	accessStr := plan.Access.ValueString()
+	// Call Models.GrantModel
 	for _, user := range users {
-		err := client.Models.GrantModel(juju.GrantModelInput{
+		err := a.client.Models.GrantModel(juju.GrantModelInput{
 			User:       user,
-			Access:     access,
+			Access:     accessStr,
 			ModelUUIDs: modelUUIDs,
 		})
 		if err != nil {
-			return diag.FromErr(err)
+			resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create access model resource, got error: %s", err))
+			return
 		}
 	}
+	plan.ID = types.StringValue(newAccessModelIDFrom(modelNameStr, accessStr, users))
 
-	d.SetId(fmt.Sprintf("%s:%s", model, access))
-
-	return diags
+	// Set the plan onto the Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
 }
 
-func resourceAccessModelRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*juju.Client)
-
-	var diags diag.Diagnostics
+func (a *accessModelResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	// Check first if the client is configured
+	if a.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "read")
+		return
+	}
+	var plan accessModelResourceModel
 
-	id := strings.Split(d.Id(), ":")
-	usersInterface := d.Get("users").([]interface{})
-	stateUsers := make([]string, len(usersInterface))
-	for i, v := range usersInterface {
-		stateUsers[i] = v.(string)
+	// Get the Terraform state from the request into the plan
+	resp.Diagnostics.Append(req.State.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
-	uuid, err := client.Models.ResolveModelUUID(id[0])
+	modelUUID, access, stateUsers := retrieveAccessModelDataFromID(&plan)
```

Comment:
Just send the ID, an entire plan is not required. Or is the function misnamed? 

Isn't the var named `modelUUID` really an `modelName`?